### PR TITLE
Set linker name when using AppleClang

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
Signed-off-by: Lloyd-Pottiger <yan1579196623@gamil.com>

### What problem does this PR solve?

Issue Number: ref #6129

Problem Summary:

### What is changed and how it works?

set linker name equal to `ld` when using AppleClang.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by JaySon-Huang** [comment]
Can you elaborate more why the warnings appears and why this can suppress the warnings?

**Review by ywqzzy — APPROVED** [review]

**Comment by Lloyd-Pottiger** [comment]
> Can you elaborate more why the warnings appears and why this can suppress the warnings?

I found an explanation: https://discourse.itk.org/t/visibility-settings/596/4

**Comment by breezewish** [comment]
Is it possible to specify this flag more precisely (by setting it as a flag to some targets), instead of globally?

**Comment by Lloyd-Pottiger** [comment]
> Is it possible to specify this flag more precisely (by setting it as a flag to some targets that set visibility incorrectly), instead of globally? The major reason drives me think about this direction is setting this flag causes everything to be rebuilt .

In my device, `ninja tiflash` output 164106 lines ld warnings each time..... almost every subdirectories......

> It would be also better to discover why only MacOS platforms are affected, while we using LLVM compiler for both Linux and MacOS.

I am working on it, but have not find the answer yet.

**Comment by Lloyd-Pottiger** [comment]
@JaySon-Huang @breezewish it is caused by the default linker `ld` in apple. If we use llvm linker `lld`, this problem can fixed. But why the `ld` report those warnings, I can not find the answer.

File: CMakeLists.txt [hunk 1]
```diff
@@ -248,17 +248,22 @@ endif()
...
 find_program (GOLD_PATH NAMES "gold")
+if (APPLE)
+    find_program (LD_PATH NAMES "ld")
+endif ()
 
-if (NOT LINKER_NAME AND NOT APPLE)
-    if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND LLD_PATH)
+if (NOT LINKER_NAME)
+    if (CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang" AND LD_PATH)
+        set (LINKER_NAME "ld")
```
Comment (breezewish) [inline CMakeLists.txt:257]: Why do we specifically use ld instead of lld for AppleClang?

Comment (Lloyd-Pottiger) [reply CMakeLists.txt:257]: use lld may cause https://github.com/pingcap/tiflash/issues/5368

**Review by breezewish — APPROVED** [review]

**Pull request merged**